src/core/lib/gprpp/windows/directory_reader.cc

```diff
@@ -60,20 +58,20 @@ std::unique_ptr<DirectoryReader> MakeDirectoryReader(
 // Reference for reading directory in Windows:
 // https://stackoverflow.com/questions/612097/how-can-i-get-the-list-of-files-in-a-directory-using-c-or-c
 // https://learn.microsoft.com/en-us/windows/win32/fileio/listing-the-files-in-a-directory
-absl::StatusOr<std::vector<std::string>>
-DirectoryReaderImpl::GetDirectoryContents() {
-  std::string search_path = absl::StrCat(directory_path_, "/*.*");
-  std::vector<std::string> files;
+absl::Status DirectoryReaderImpl::ForEach(
+    absl::FunctionRef<void(absl::string_view)> callback) {
+  // std::string search_path = absl::StrCat(directory_path_, "/*.*");
+  std::string search_path = absl::StrCat(directory_path_, "/*");
   WIN32_FIND_DATA find_data;
   HANDLE hFind = ::FindFirstFile(search_path.c_str(), &find_data);
   if (hFind != INVALID_HANDLE_VALUE) {
     do {
       if (!(find_data.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY)) {
-        files.push_back(std::string(find_data.cFileName));
+        callback(std::string(find_data.cFileName));
       }
     } while (::FindNextFile(hFind, &find_data));
     ::FindClose(hFind);
-    return files;
+    return absl::OkStatus();
   } else {
     return absl::InternalError("Could not read crl directory.");
```

Comment:
As a general rule, please keep the code that handles an error next to the condition used to detect the error.  In other words, instead of this:

```
if (!error_condition) {
  // ...long block to handle success case...
} else {
  HandleError();
}
```

Please say this:

```
if (error_condition) {
  HandleError();
} else {
  // ...long block to handle success case...
}
```

Or better yet, in cases like this one, you can return immediately in the error case, thus eliminating the need for indentation in the success case:

```
if (error_condition) {
  return error;
}
// ...long block to handle success case...
```

Reply:
Made the pattern better